File: src/core/message_loop.ts
import type { ChatMessage, ToolCall, ToolCallDelta } from "../api/types";
import { OpenRouterClient } from "../api/client";
import { getToolDefinitions } from "../tools/index";
import { buildSystemPrompt } from "./system_prompt";
import { executeToolCalls, type ToolExecutionResult } from "./tool_executor";

export interface MessageLoopCallbacks {
  onToken: (token: string) => void;
  onToolStart: (toolCalls: ToolCall[]) => void;
  onToolResult: (results: ToolExecutionResult[]) => void;
  onComplete: (text: string) => void;
  onError: (error: Error) => void;
}

export async function runMessageLoop(
  client: OpenRouterClient,
  messages: ChatMessage[],
  model: string,
  callbacks: MessageLoopCallbacks,
): Promise<ChatMessage[]> {
  const systemMessage: ChatMessage = {
    role: "system",
    content: buildSystemPrompt(),
  };

  const tools = getToolDefinitions();
  let continueLoop = true;

  while (continueLoop) {
    try {
      let assistantText = "";
      const toolCallDeltas: Map<
        number,
        { id: string; name: string; arguments: string }
      > = new Map();

      const stream = client.streamChatCompletion({
        model,
        messages: [systemMessage, ...messages],
        tools: tools.length > 0 ? tools : undefined,
      });

      let finishReason: string | null = null;

      for await (const chunk of stream) {
        const choice = chunk.choices[0];
        if (!choice) continue;

        // Handle text content
        if (choice.delta.content) {
          assistantText += choice.delta.content;
          callbacks.onToken(choice.delta.content);
        }

        // Handle tool call deltas
        if (choice.delta.tool_calls) {
          for (const delta of choice.delta.tool_calls) {
            accumulateToolCallDelta(toolCallDeltas, delta);
          }
        }

        if (choice.finish_reason) {
          finishReason = choice.finish_reason;
        }
      }

      // Build the completed tool calls array
      const toolCalls: ToolCall[] = Array.from(toolCallDeltas.values()).map(
        (tc) => ({
          id: tc.id,
          type: "function" as const,
          function: {
            name: tc.name,
            arguments: tc.arguments,
          },
        }),
      );

      if (
        finishReason === "tool_calls" ||
        (toolCalls.length > 0 && finishReason !== "stop")
      ) {
        // Add assistant message with tool calls
        const assistantMessage: ChatMessage = {
          role: "assistant",
          content: assistantText || null,
          tool_calls: toolCalls,
        };
        messages.push(assistantMessage);

        callbacks.onToolStart(toolCalls);

        // Execute tools
        const { messages: toolMessages, results } =
          await executeToolCalls(toolCalls);

        messages.push(...toolMessages);
        callbacks.onToolResult(results);

        // Continue the loop for another round
      } else {
        // finish_reason is "stop" or no tool calls - we're done
        if (assistantText) {
          messages.push({
            role: "assistant",
            content: assistantText,
          });
        }
        callbacks.onComplete(assistantText);
        continueLoop = false;
      }
    } catch (error) {
      callbacks.onError(
        error instanceof Error ? error : new Error(String(error)),
      );
      continueLoop = false;
    }
  }

  return messages;
}

function accumulateToolCallDelta(
  accumulator: Map<number, { id: string; name: string; arguments: string }>,
  delta: ToolCallDelta,
) {
  const existing = accumulator.get(delta.index);

  if (!existing) {
    accumulator.set(delta.index, {
      id: delta.id || "",
      name: delta.function?.name || "",
      arguments: delta.function?.arguments || "",
    });
  } else {
    if (delta.id) existing.id = delta.id;
    if (delta.function?.name) existing.name = delta.function.name;
    if (delta.function?.arguments)
      existing.arguments += delta.function.arguments;
  }
}

File: internal/core/message_loop.go
package core

import (
	"context"

	"github.com/sagnikc395/kai/internal/api"
	"github.com/sagnikc395/kai/internal/tools"
)

type MessageLoopCallbacks struct {
	OnToken      func(token string)
	OnToolStart  func(toolCalls []api.ToolCall)
	OnToolResult func(results []ToolExecutionResult)
	OnComplete   func(text string)
	OnError      func(err error)
}

type toolCallAccumulator struct {
	id        string
	name      string
	arguments string
}

func RunMessageLoop(ctx context.Context, client *api.OpenRouterClient, messages []api.ChatMessage, model string, callbacks MessageLoopCallbacks) []api.ChatMessage {
	systemMessage := api.ChatMessage{
		Role:    "system",
		Content: api.StringPtr(BuildSystemPrompt()),
	}

	toolDefinitions := tools.Definitions()
	continueLoop := true

	for continueLoop {
		assistantText := ""
		toolCallDeltas := map[int]*toolCallAccumulator{}
		finishReason := ""

		requestMessages := make([]api.ChatMessage, 0, len(messages)+1)
		requestMessages = append(requestMessages, systemMessage)
		requestMessages = append(requestMessages, messages...)

		err := client.StreamChatCompletion(ctx, api.ChatCompletionRequest{
			Model:    model,
			Messages: requestMessages,
			Tools:    toolDefinitions,
		}, func(chunk api.ChatCompletionChunk) error {
			if len(chunk.Choices) == 0 {
				return nil
			}

			choice := chunk.Choices[0]
			if choice.Delta.Content != nil && *choice.Delta.Content != "" {
				assistantText += *choice.Delta.Content
				if callbacks.OnToken != nil {
					callbacks.OnToken(*choice.Delta.Content)
				}
			}

			for _, delta := range choice.Delta.ToolCalls {
				accumulateToolCallDelta(toolCallDeltas, delta)
			}

			if choice.FinishReason != "" {
				finishReason = choice.FinishReason
			}
			return nil
		})

		if err != nil {
			if callbacks.OnError != nil {
				callbacks.OnError(err)
			}
			return messages
		}

		toolCalls := buildToolCalls(toolCallDeltas)
		if finishReason == "tool_calls" || (len(toolCalls) > 0 && finishReason != "stop") {
			messages = append(messages, api.ChatMessage{
				Role:      "assistant",
				Content:   nullableContent(assistantText),
				ToolCalls: toolCalls,
			})

			if callbacks.OnToolStart != nil {
				callbacks.OnToolStart(toolCalls)
			}

			toolMessages, results := ExecuteToolCalls(toolCalls)
			messages = append(messages, toolMessages...)

			if callbacks.OnToolResult != nil {
				callbacks.OnToolResult(results)
			}
			continue
		}

		if assistantText != "" {
			messages = append(messages, api.ChatMessage{
				Role:    "assistant",
				Content: api.StringPtr(assistantText),
			})
		}
		if callbacks.OnComplete != nil {
			callbacks.OnComplete(assistantText)
		}
		continueLoop = false
	}

	return messages
}

func accumulateToolCallDelta(accumulator map[int]*toolCallAccumulator, delta api.ToolCallDelta) {
	existing, ok := accumulator[delta.Index]
	if !ok {
		existing = &toolCallAccumulator{}
		accumulator[delta.Index] = existing
	}
	if delta.ID != "" {
		existing.id = delta.ID
	}
	if delta.Function != nil {
		if delta.Function.Name != "" {
			existing.name = delta.Function.Name
		}
		if delta.Function.Arguments != "" {
			existing.arguments += delta.Function.Arguments
		}
	}
}

func buildToolCalls(accumulator map[int]*toolCallAccumulator) []api.ToolCall {
	toolCalls := make([]api.ToolCall, 0, len(accumulator))
	for index := 0; index < len(accumulator); index++ {
		accumulated, ok := accumulator[index]
		if !ok {
			continue
		}
		toolCalls = append(toolCalls, api.ToolCall{
			ID:   accumulated.id,
			Type: "function",
			Function: api.FunctionCall{
				Name:      accumulated.name,
				Arguments: accumulated.arguments,
			},
		})
	}
	return toolCalls
}

func nullableContent(text string) *string {
	if text == "" {
		return nil
	}
	return api.StringPtr(text)
}
